gluefactory/utils/image.py

```diff
@@ -128,3 +131,94 @@ def numpy_image_to_torch(image: np.ndarray) -> torch.Tensor:
 def load_image(path: Path, grayscale=False) -> torch.Tensor:
     image = read_image(path, grayscale=grayscale)
     return numpy_image_to_torch(image)
+
+
+def grid_sample(
+    image: torch.Tensor, coords: torch.Tensor, interpolation: str = "bilinear"
+):
+    assert image.dim() == coords.dim()
+    is_batched = image.dim() == 4
+
+    if is_batched:
+        assert coords.dim() == 4
+        return F.grid_sample(
+            image.to(coords.device), coords, interpolation, align_corners=False
+        )
+    else:
+        return F.grid_sample(
+            image[None].to(coords.device),
+            coords[None],
+            mode=interpolation,
+            align_corners=False,
+        )[0]
+
+
+def get_pixel_grid(
+    *,
+    fmap: torch.Tensor | None = None,  # B x H X W X D
+    camera: Camera | None = None,
+    size: Optional[tuple[int, int]] = None,
+    device: torch.device | None = None,
+    dtype=torch.float32,
+    normalized: bool = False,
+) -> torch.Tensor:
+    if fmap is None:
+        if camera is None:
+            if size is None:
+                raise ValueError("Specify fmap, size, or camera")
+            w, h = size
+        else:
+            w, h = camera.size.int()
+            device = camera.device
+            dtype = camera.dtype
+    else:
+        *_, h, w, _ = fmap.shape
+        device = fmap.device
+        dtype = fmap.dtype
+    grid = torch.stack(
+        torch.meshgrid(
+            torch.arange(w, dtype=dtype, device=device),
+            torch.arange(h, dtype=dtype, device=device),
+            indexing="xy",
+        ),
+        dim=-1,
+    )
+    if fmap.ndim == 4:
+        b = fmap.shape[0]
+        grid = grid[None].expand(b, -1, -1, -1)
+    grid += 0.5
+    if normalized:
+        grid *= 2 / grid.new_tensor([w, h])
+        grid -= 1
+    elif fmap is not None and camera is not None:
+        # In case fmaps are at a different image resolution
+        grid *= camera.size / grid.new_tensor([w, h])
+    return grid
+
+
+def coords_to_image(coords):
```

Comment:
Maybe add one-line docstrings explaining what these do and the expected input and output shapes?